execution/gethexec/express_lane_service.go

```
@@ -0,0 +1,432 @@
+			// 				0,
+			// 				make(map[uint64]*msgAndResult),
+			// 			})
+			// 		}
+			// 		es.roundInfoMutex.Unlock()
+			// 	}
+			// }
+			fromBlock = toBlock + 1
+		}
+	})
+}
+
+func (es *expressLaneService) currentRoundHasController() bool {
+	controller, ok := es.roundControl.Load(es.roundTimingInfo.RoundNumber())
+	if !ok {
+		return false
+	}
+	return controller != (common.Address{})
+}
+
+// sequenceExpressLaneSubmission with the roundInfo lock held, validates sequence number and sender address fields of the message
+// adds the message to the transaction queue and waits for the response
+func (es *expressLaneService) sequenceExpressLaneSubmission(
+	ctx context.Context,
+	msg *timeboost.ExpressLaneSubmission,
+) error {
+	unlockByDefer := true
+	es.roundInfoMutex.Lock()
+	defer func() {
+		if unlockByDefer {
+			es.roundInfoMutex.Unlock()
+		}
+	}()
+
+	// Below code block isn't a repetition, it prevents stale messages to be accepted during control transfer within or after the round ends!
+	controller, ok := es.roundControl.Load(msg.Round)
+	if !ok {
+		return timeboost.ErrNoOnchainController
+	}
+	sender, err := msg.Sender() // Doesn't recompute sender address
+	if err != nil {
+		return err
+	}
+	if sender != controller {
+		return timeboost.ErrNotExpressLaneController
+	}
+
+	// If expressLaneRoundInfo for current round doesn't exist yet, we'll add it to the cache
+	if !es.roundInfo.Contains(msg.Round) {
+		es.roundInfo.Add(msg.Round, &expressLaneRoundInfo{
+			0,
+			make(map[uint64]*msgAndResult),
+		})
+	}
+	roundInfo, _ := es.roundInfo.Get(msg.Round)
+
+	// Check if the submission nonce is too low.
+	if msg.SequenceNumber < roundInfo.sequence {
+		return timeboost.ErrSequenceNumberTooLow
+	}
+
+	// Check if a duplicate submission exists already, and reject if so.
+	if _, exists := roundInfo.msgAndResultBySequenceNumber[msg.SequenceNumber]; exists {
+		return timeboost.ErrDuplicateSequenceNumber
+	}
```

in these two we're handling an already-existing sequence number. I think we don't want to return an error if it's the same message. Compare the new message to the existing one by comparing signatures. If they are the same - return nil.

---

added this